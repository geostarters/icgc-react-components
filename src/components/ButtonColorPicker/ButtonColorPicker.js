/* eslint-disable no-invalid-this */
// @flow

import React from "react";
import PropTypes from "prop-types";
import { ChromePicker } from "react-color";

import styles from "./ButtonColorPicker.module.css";

export default class ButtonColorPicker extends React.PureComponent {

	state = {
		displayColorPicker: false,
		color: this.props.color
	};

	handleClick = () => this.setState({ displayColorPicker: !this.state.displayColorPicker });

	handleClose = () => this.setState({ displayColorPicker: false });

	handleChange = (color) => {

		this.setState(prevState => ({
			...prevState,
			color
		}));
		this.props.handleChangeColor(color, this.props.index);

	}

	render() {

		const { color } = this.state;

		return (
			<div className={ styles.buttonContainer }>
				<div className={ styles.swatch } onClick={ this.handleClick }>
					<div className={ styles.color } style={{background: `rgba(${color.rgb.r}, ${color.rgb.g}, ${color.rgb.b}, ${color.rgb.a})`}}/>
				</div>
				{ this.state.displayColorPicker ? <div className={ styles.popover }>
					<div className={ styles.cover } onClick={ this.handleClose }/>
					<ChromePicker color={ color.rgb } onChange={ this.handleChange } />
				</div> : null }

			</div>
		);

	}

}

ButtonColorPicker.propTypes = {
	handleChangeColor: PropTypes.func,
	color: PropTypes.object,
	index: PropTypes.oneOfType([PropTypes.number, PropTypes.string])
};

ButtonColorPicker.defaultProps = {
	handleChangeColor: () => {},
	color: {rgb: {r: 255, g:0, b: 0, a: 1} },
	index: 0
};
