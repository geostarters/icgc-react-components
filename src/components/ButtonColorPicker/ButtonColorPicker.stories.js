//@flow

import React from "react";
import { action } from "@storybook/addon-actions";

import ButtonColorPicker from "./ButtonColorPicker";

import "semantic-ui-css/semantic.min.css";

export default {
	title: "ButtonColorPicker",
	component: ButtonColorPicker
};

const Template = args => (<ButtonColorPicker {...args}/>);

export const _ButtonColorPicker = Template.bind({});
_ButtonColorPicker.storyName = "ButtonColorPicker";

export const ButtonColorPickerColor = Template.bind({});
ButtonColorPickerColor.storyName = "ButtonColorPicker initialColor";
ButtonColorPickerColor.args = {
	initialColor: {rgb: {r: 255, g:255, b: 0, a: 1} },
	handleChangeColor: action("onChange")
};
