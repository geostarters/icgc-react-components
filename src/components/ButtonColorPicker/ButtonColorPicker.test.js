// @flow

import React from "react";
import ButtonColorPicker from "./ButtonColorPicker";
import  { shallow } from "enzyme";
import toJSON from "enzyme-to-json";
import { ChromePicker } from "react-color";

describe("ButtonColorPicker Component", () => {

	const buttonColorPickerProps = {
		initialColor: {rgb: {r: 255, g:255, b: 0, a: 1} },
		index: 0,
		handleChangeColor: jest.fn()
	};
	const component = shallow(<ButtonColorPicker {...buttonColorPickerProps}></ButtonColorPicker>);

	it("renders correct properties", () => {

		expect(toJSON(component)).toMatchSnapshot();

	});

	it("has a element with  swatch class", () => {

		expect(component.find(".swatch")).toHaveLength(1);

	});

	it("has the correct color in the element", () => {

		const { initialColor: color } = buttonColorPickerProps;
		expect(component.find(".swatch").find(".color").get(0).props.style).toHaveProperty("background", `rgba(${color.rgb.r}, ${color.rgb.g}, ${color.rgb.b}, ${color.rgb.a})`);

	});

	it("shows ColorPicker after click", () => {

		component.find(".swatch").simulate("click");
		expect(component.state("displayColorPicker")).toBeTruthy();

	});

	it("call handle function after changing color", () => {

		component.find(ChromePicker).simulate("change", {rgb: {r: 255, g:255, b: 0, a: 1}});
		expect(buttonColorPickerProps.handleChangeColor).toBeCalled();

	});

});
