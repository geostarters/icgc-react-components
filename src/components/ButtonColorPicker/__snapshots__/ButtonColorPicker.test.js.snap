// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonColorPicker Component renders correct properties 1`] = `
<div
  className="buttonContainer"
>
  <div
    className="swatch"
    onClick={[Function]}
  >
    <div
      className="color"
      style={
        Object {
          "background": "rgba(255, 255, 0, 1)",
        }
      }
    />
  </div>
</div>
`;
